Reduce JWT error logging in WebSockets

Currently, WebSocket connectors pollute the log with repeated exceptions regarding expired JWTs. This is quite annoying as it is difficult to navigate the log to get to relevant information.

The following exception repeats in the log:

```
org.springframework.security.authentication.AuthenticationServiceException: JWT expired at 2024-09-24T15:36:49Z. Current time: 2024-10-02T13:57:57Z, a difference of 685268282 milliseconds.  Allowed clock skew: 0 milliseconds.
dss-test_termit-server_1  | 	at org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationProvider.getJwt(JwtAuthenticationProvider.java:106)
dss-test_termit-server_1  | 	at org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationProvider.authenticate(JwtAuthenticationProvider.java:88)
dss-test_termit-server_1  | 	at cz.cvut.kbss.termit.security.WebSocketJwtAuthorizationInterceptor.process(WebSocketJwtAuthorizationInterceptor.java:71)
dss-test_termit-server_1  | 	at cz.cvut.kbss.termit.security.WebSocketJwtAuthorizationInterceptor.preSend(WebSocketJwtAuthorizationInterceptor.java:45)
dss-test_termit-server_1  | 	at org.springframework.messaging.support.AbstractMessageChannel$ChannelInterceptorChain.applyPreSend(AbstractMessageChannel.java:181)
dss-test_termit-server_1  | 	at org.springframework.messaging.support.AbstractMessageChannel.send(AbstractMessageChannel.java:135)
dss-test_termit-server_1  | 	at org.springframework.messaging.support.AbstractMessageChannel.send(AbstractMessageChannel.java:125)
dss-test_termit-server_1  | 	at org.springframework.web.socket.messaging.StompSubProtocolHandler.handleMessageFromClient(StompSubProtocolHandler.java:343)
dss-test_termit-server_1  | 	at org.springframework.web.socket.messaging.SubProtocolWebSocketHandler.handleMessage(SubProtocolWebSocketHandler.java:356)
dss-test_termit-server_1  | 	at org.springframework.web.socket.handler.WebSocketHandlerDecorator.handleMessage(WebSocketHandlerDecorator.java:75)
dss-test_termit-server_1  | 	at org.springframework.web.socket.handler.LoggingWebSocketHandlerDecorator.handleMessage(LoggingWebSocketHandlerDecorator.java:56)
dss-test_termit-server_1  | 	at org.springframework.web.socket.handler.ExceptionWebSocketHandlerDecorator.handleMessage(ExceptionWebSocketHandlerDecorator.java:58)
dss-test_termit-server_1  | 	at org.springframework.web.socket.adapter.standard.StandardWebSocketHandlerAdapter.handleTextMessage(StandardWebSocketHandlerAdapter.java:113)
dss-test_termit-server_1  | 	at org.springframework.web.socket.adapter.standard.StandardWebSocketHandlerAdapter$3.onMessage(StandardWebSocketHandlerAdapter.java:84)
dss-test_termit-server_1  | 	at org.springframework.web.socket.adapter.standard.StandardWebSocketHandlerAdapter$3.onMessage(StandardWebSocketHandlerAdapter.java:81)
dss-test_termit-server_1  | 	at org.apache.tomcat.websocket.WsFrameBase.sendMessageText(WsFrameBase.java:390)
dss-test_termit-server_1  | 	at org.apache.tomcat.websocket.server.WsFrameServer.sendMessageText(WsFrameServer.java:130)
dss-test_termit-server_1  | 	at org.apache.tomcat.websocket.WsFrameBase.processDataText(WsFrameBase.java:484)
dss-test_termit-server_1  | 	at org.apache.tomcat.websocket.WsFrameBase.processData(WsFrameBase.java:284)
dss-test_termit-server_1  | 	at org.apache.tomcat.websocket.WsFrameBase.processInputBuffer(WsFrameBase.java:130)
dss-test_termit-server_1  | 	at org.apache.tomcat.websocket.server.WsFrameServer.onDataAvailable(WsFrameServer.java:85)
dss-test_termit-server_1  | 	at org.apache.tomcat.websocket.server.WsFrameServer.doOnDataAvailable(WsFrameServer.java:184)
dss-test_termit-server_1  | 	at org.apache.tomcat.websocket.server.WsFrameServer.notifyDataAvailable(WsFrameServer.java:164)
dss-test_termit-server_1  | 	at org.apache.tomcat.websocket.server.WsHttpUpgradeHandler.upgradeDispatch(WsHttpUpgradeHandler.java:152)
dss-test_termit-server_1  | 	at org.apache.coyote.http11.upgrade.UpgradeProcessorInternal.dispatch(UpgradeProcessorInternal.java:60)
dss-test_termit-server_1  | 	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:57)
dss-test_termit-server_1  | 	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:904)
dss-test_termit-server_1  | 	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1741)
dss-test_termit-server_1  | 	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52)
dss-test_termit-server_1  | 	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1190)
dss-test_termit-server_1  | 	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
dss-test_termit-server_1  | 	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:63)
dss-test_termit-server_1  | 	at java.base/java.lang.Thread.run(Thread.java:840)
dss-test_termit-server_1  | Caused by: org.springframework.security.oauth2.jwt.JwtException: JWT expired at 2024-09-24T15:36:49Z. Current time: 2024-10-02T13:57:57Z, a difference of 685268282 milliseconds.  Allowed clock skew: 0 milliseconds.
dss-test_termit-server_1  | 	at cz.cvut.kbss.termit.security.TermitJwtDecoder.decode(TermitJwtDecoder.java:57)
dss-test_termit-server_1  | 	at org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationProvider.getJwt(JwtAuthenticationProvider.java:99)
dss-test_termit-server_1  | 	... 32 common frames omitted
dss-test_termit-server_1  | Caused by: cz.cvut.kbss.termit.exception.TokenExpiredException: JWT expired at 2024-09-24T15:36:49Z. Current time: 2024-10-02T13:57:57Z, a difference of 685268282 milliseconds.  Allowed clock skew: 0 milliseconds.
dss-test_termit-server_1  | 	at cz.cvut.kbss.termit.security.JwtUtils.getClaimsFromToken(JwtUtils.java:150)
dss-test_termit-server_1  | 	at cz.cvut.kbss.termit.security.TermitJwtDecoder.decode(TermitJwtDecoder.java:38)
dss-test_termit-server_1  | 	... 33 common frames omitted

```

--- a/src/main/java/cz/cvut/kbss/termit/websocket/handler/StompExceptionHandler.java
+++ b/src/main/java/cz/cvut/kbss/termit/websocket/handler/StompExceptionHandler.java
@@ -1,6 +1,7 @@
 package cz.cvut.kbss.termit.websocket.handler;
 
 import jakarta.annotation.Nonnull;
+import jakarta.annotation.Nullable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.messaging.Message;
@@ -23,10 +24,15 @@ public class StompExceptionHandler extends StompSubProtocolErrorHandler {
 
     @Override
     protected @Nonnull Message<byte[]> handleInternal(@Nonnull StompHeaderAccessor errorHeaderAccessor,
-                                                      @Nonnull byte[] errorPayload, Throwable cause,
-                                                      StompHeaderAccessor clientHeaderAccessor) {
+                                                      @Nonnull byte[] errorPayload,
+                                                      @Nullable Throwable cause,
+                                                      @Nullable StompHeaderAccessor clientHeaderAccessor) {
         final Message<?> message = MessageBuilder.withPayload(errorPayload).setHeaders(errorHeaderAccessor).build();
-        final boolean handled = webSocketExceptionHandler.delegate(message, cause);
+        Throwable causeToHandle = cause;
+        if (causeToHandle != null && causeToHandle.getCause() != null) {
+            causeToHandle = causeToHandle.getCause();
+        }
+        final boolean handled = webSocketExceptionHandler.delegate(message, causeToHandle);
 
         if (!handled) {
             LOG.error("STOMP sub-protocol exception", cause);

--- a/src/main/java/cz/cvut/kbss/termit/websocket/handler/StompExceptionHandler.java
+++ b/src/main/java/cz/cvut/kbss/termit/websocket/handler/StompExceptionHandler.java
@@ -8,9 +8,6 @@ import org.springframework.messaging.simp.stomp.StompHeaderAccessor;
 import org.springframework.messaging.support.MessageBuilder;
 import org.springframework.web.socket.messaging.StompSubProtocolErrorHandler;
 
-import java.lang.reflect.InvocationTargetException;
-import java.lang.reflect.Method;
-
 /**
  * calls {@link WebSocketExceptionHandler} when possible, otherwise logs exception as error
  */
@@ -29,51 +26,12 @@ public class StompExceptionHandler extends StompSubProtocolErrorHandler {
                                                       @Nonnull byte[] errorPayload, Throwable cause,
                                                       StompHeaderAccessor clientHeaderAccessor) {
         final Message<?> message = MessageBuilder.withPayload(errorPayload).setHeaders(errorHeaderAccessor).build();
-        boolean handled = false;
-        try {
-            handled = delegate(message, cause);
-        } catch (InvocationTargetException e) {
-            LOG.error("Exception thrown during exception handler invocation", e);
-        } catch (IllegalAccessException unexpected) {
-            // is checked by delegate
-        }
+        final boolean handled = webSocketExceptionHandler.delegate(message, cause);
 
         if (!handled) {
             LOG.error("STOMP sub-protocol exception", cause);
         }
 
         return super.handleInternal(errorHeaderAccessor, errorPayload, cause, clientHeaderAccessor);
-    }
-
-    /**
-     * Tries to match method on {@link #webSocketExceptionHandler}
-     *
-     * @return true when a method was found and called, false otherwise
-     * @throws IllegalArgumentException never
-     */
-    private boolean delegate(Message<?> message, Throwable throwable)
-            throws InvocationTargetException, IllegalAccessException {
-        if (throwable instanceof Exception exception) {
-            Method[] methods = webSocketExceptionHandler.getClass().getMethods();
-            for (final Method method : methods) {
-                if (!method.canAccess(webSocketExceptionHandler)) {
-                    continue;
-                }
-                Class<?>[] params = method.getParameterTypes();
-                if (params.length != 2) {
-                    continue;
-                }
-                if (params[0].isAssignableFrom(message.getClass()) && params[1].isAssignableFrom(exception.getClass())) {
-                    // message, exception
-                    method.invoke(webSocketExceptionHandler, message, exception);
-                    return true;
-                } else if (params[0].isAssignableFrom(exception.getClass()) && params[1].isAssignableFrom(message.getClass())) {
-                    // exception, message
-                    method.invoke(webSocketExceptionHandler, exception, message);
-                    return true;
-                }
-            }
-        }
-        return false;
     }
 }

--- a/src/main/java/cz/cvut/kbss/termit/websocket/handler/WebSocketExceptionHandler.java
+++ b/src/main/java/cz/cvut/kbss/termit/websocket/handler/WebSocketExceptionHandler.java
@@ -11,6 +11,7 @@ import cz.cvut.kbss.termit.exception.InvalidLanguageConstantException;
 import cz.cvut.kbss.termit.exception.InvalidParameterException;
 import cz.cvut.kbss.termit.exception.InvalidPasswordChangeRequestException;
 import cz.cvut.kbss.termit.exception.InvalidTermStateException;
+import cz.cvut.kbss.termit.exception.JwtException;
 import cz.cvut.kbss.termit.exception.NotFoundException;
 import cz.cvut.kbss.termit.exception.PersistenceException;
 import cz.cvut.kbss.termit.exception.ResourceExistsException;
@@ -25,6 +26,7 @@ import cz.cvut.kbss.termit.exception.WebServiceIntegrationException;
 import cz.cvut.kbss.termit.exception.importing.UnsupportedImportMediaTypeException;
 import cz.cvut.kbss.termit.exception.importing.VocabularyImportException;
 import cz.cvut.kbss.termit.rest.handler.ErrorInfo;
+import cz.cvut.kbss.termit.util.ExceptionUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.messaging.Message;
@@ -33,12 +35,15 @@ import org.springframework.messaging.handler.annotation.MessageExceptionHandler;
 import org.springframework.messaging.simp.annotation.SendToUser;
 import org.springframework.messaging.simp.stomp.StompHeaderAccessor;
 import org.springframework.security.access.AccessDeniedException;
+import org.springframework.security.authentication.AuthenticationServiceException;
 import org.springframework.security.core.AuthenticationException;
 import org.springframework.security.core.userdetails.UsernameNotFoundException;
 import org.springframework.web.bind.annotation.ControllerAdvice;
 import org.springframework.web.context.request.async.AsyncRequestNotUsableException;
 import org.springframework.web.multipart.MaxUploadSizeExceededException;
 
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
 import java.net.URISyntaxException;
 
 import static cz.cvut.kbss.termit.util.ExceptionUtils.findCause;
@@ -92,8 +97,61 @@ public class WebSocketExceptionHandler {
                                                        e.getParameters());
     }
 
+    /**
+     * Tries to match method on this object that matches signature with params
+     *
+     * @return true when a method was found and called, false otherwise
+     * @throws IllegalArgumentException never
+     */
+    public boolean delegate(Message<?> message, Throwable throwable) {
+        try {
+            return delegateInternal(message, throwable.getCause());
+        } catch (InvocationTargetException invEx) {
+            LOG.error("Exception thrown during exception handler invocation", invEx);
+        } catch (IllegalAccessException unexpected) {
+            // is checked by delegate
+        }
+        return false;
+    }
+
+    /**
+     * Tries to match method on this object that matches signature with params
+     *
+     * @return true when a method was found and called, false otherwise
+     * @throws IllegalArgumentException never
+     */
+    private boolean delegateInternal(Message<?> message, Throwable throwable)
+            throws InvocationTargetException, IllegalAccessException {
+        if (throwable instanceof Exception exception) {
+            Method[] methods = this.getClass().getMethods();
+            for (final Method method : methods) {
+                if (!method.canAccess(this) || method.getName().equals("delegate") || method.getName().equals("delegateInternal")) {
+                    continue;
+                }
+                Class<?>[] params = method.getParameterTypes();
+                if (params.length != 2) {
+                    continue;
+                }
+                if (params[0].isAssignableFrom(message.getClass()) && params[1].isAssignableFrom(exception.getClass())) {
+                    // message, exception
+                    method.invoke(this, message, exception);
+                    return true;
+                } else if (params[0].isAssignableFrom(exception.getClass()) && params[1].isAssignableFrom(message.getClass())) {
+                    // exception, message
+                    method.invoke(this, exception, message);
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
     @MessageExceptionHandler
     public void messageDeliveryException(Message<?> message, MessageDeliveryException e) {
+        if (!(e.getCause() instanceof MessageDeliveryException) && delegate(message, e.getCause())) {
+            return;
+        }
+
         // messages without destination will be logged only on trace
         (hasDestination(message) ? LOG.atError() : LOG.atTrace())
                 .setMessage("Failed to send message with destination {}: {}")
@@ -144,11 +202,16 @@ public class WebSocketExceptionHandler {
         return errorInfo(message, e);
     }
 
-    @MessageExceptionHandler(AuthenticationException.class)
+    @MessageExceptionHandler({AuthenticationException.class, AuthenticationServiceException.class})
     public ErrorInfo authenticationException(Message<?> message, AuthenticationException e) {
-        LOG.atDebug().setCause(e).log(e.getMessage());
-        LOG.atError().setMessage("Authentication failure during message processing: {}\nMessage: {}")
+        LOG.atWarn().setMessage("Authentication failure during message processing: {}\nMessage: {}")
            .addArgument(e.getMessage()).addArgument(message::toString).log();
+
+        if (ExceptionUtils.findCause(e, JwtException.class).isPresent()) {
+            return errorInfo(message, e);
+        }
+
+        LOG.atDebug().setCause(e).log(e.getMessage());
         return errorInfo(message, e);
     }
 

--- a/src/main/java/cz/cvut/kbss/termit/websocket/handler/WebSocketExceptionHandler.java
+++ b/src/main/java/cz/cvut/kbss/termit/websocket/handler/WebSocketExceptionHandler.java
@@ -45,11 +45,15 @@ import org.springframework.web.multipart.MaxUploadSizeExceededException;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.net.URISyntaxException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
 
 import static cz.cvut.kbss.termit.util.ExceptionUtils.findCause;
 
 /**
- * @implSpec Should reflect {@link cz.cvut.kbss.termit.rest.handler.RestExceptionHandler}
+ * @implSpec Should reflect {@link cz.cvut.kbss.termit.rest.handler.RestExceptionHandler}.<br>
+ * In order for the delegation to work, the method signature of MessageExceptionHandler methods must be {@code (Message<?>, Exception)}
  */
 @SendToUser
 @ControllerAdvice
@@ -98,51 +102,73 @@ public class WebSocketExceptionHandler {
     }
 
     /**
-     * Tries to match method on this object that matches signature with params
+     * Searches available methods annotated with {@link MessageExceptionHandler} in this class
+     * when the method signature matches {@code (Message<?>, Exception)}
+     * and the exception parameter is assignable from the supplied throwable
+     * the method is called.
      *
+     * @param message the associated message
+     * @param throwable the exception to handle
      * @return true when a method was found and called, false otherwise
-     * @throws IllegalArgumentException never
      */
     public boolean delegate(Message<?> message, Throwable throwable) {
         try {
-            return delegateInternal(message, throwable.getCause());
+            return delegateInternal(message, throwable);
         } catch (InvocationTargetException invEx) {
+            // Exception handler method threw an exception
             LOG.error("Exception thrown during exception handler invocation", invEx);
         } catch (IllegalAccessException unexpected) {
-            // is checked by delegate
+            // is checked by delegateInternal
         }
         return false;
     }
 
     /**
-     * Tries to match method on this object that matches signature with params
+     * Searches available methods annotated with {@link MessageExceptionHandler} in this class
+     * when the method signature matches {@code (Message<?>, Exception)}
+     * and the exception parameter is assignable from the supplied throwable
+     * the method is called.
      *
+     * @param message the associated message
+     * @param throwable the exception to handle
      * @return true when a method was found and called, false otherwise
      * @throws IllegalArgumentException never
+     * @throws IllegalAccessException never
+     * @throws InvocationTargetException when the exception handler method throws an exception
      */
     private boolean delegateInternal(Message<?> message, Throwable throwable)
             throws InvocationTargetException, IllegalAccessException {
+        // handle only exceptions
         if (throwable instanceof Exception exception) {
-            Method[] methods = this.getClass().getMethods();
+            // find all methods annotated with MessageExceptionHandler
+            List<Method> methods = Arrays.stream(this.getClass().getMethods())
+                                         .filter(m -> m.isAnnotationPresent(MessageExceptionHandler.class)).toList();
             for (final Method method : methods) {
-                if (!method.canAccess(this) || method.getName().equals("delegate") || method.getName().equals("delegateInternal")) {
+                // check for reflection access to prevent IllegalAccessException
+                if (!method.canAccess(this)) {
                     continue;
                 }
+                // we are interested only in methods with exactly two parameters (message, exception)
                 Class<?>[] params = method.getParameterTypes();
                 if (params.length != 2) {
                     continue;
                 }
+                // check if the MessageExceptionHandler annotation has value with allowed exceptions
+                Class<? extends Throwable>[] allowedExceptions = Optional.ofNullable(method.getAnnotation(MessageExceptionHandler.class))
+                                                                         .map(MessageExceptionHandler::value).orElseGet(() -> new Class[0]);
+                // if the exception is not allowed by the annotation, skip the method
+                if (allowedExceptions.length > 0 && Arrays.stream(allowedExceptions).noneMatch(e -> e.isAssignableFrom(exception.getClass()))) {
+                    continue;
+                }
+                // validate the method signature
                 if (params[0].isAssignableFrom(message.getClass()) && params[1].isAssignableFrom(exception.getClass())) {
-                    // message, exception
+                    // call the method with message, exception parameters
                     method.invoke(this, message, exception);
-                    return true;
-                } else if (params[0].isAssignableFrom(exception.getClass()) && params[1].isAssignableFrom(message.getClass())) {
-                    // exception, message
-                    method.invoke(this, exception, message);
-                    return true;
+                    return true; // exception was handled
                 }
             }
         }
+        // throwable is not an exception or no suitable method was found
         return false;
     }
 

--- a/src/test/java/cz/cvut/kbss/termit/websocket/IntegrationWebSocketSecurityTest.java
+++ b/src/test/java/cz/cvut/kbss/termit/websocket/IntegrationWebSocketSecurityTest.java
@@ -37,6 +37,7 @@ import java.util.stream.Stream;
 import static org.awaitility.Awaitility.await;
 import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.ArgumentMatchers.isNull;
 import static org.mockito.ArgumentMatchers.notNull;
 import static org.mockito.Mockito.verify;
 
@@ -86,7 +87,8 @@ class IntegrationWebSocketSecurityTest extends BaseWebSocketIntegrationTestRunne
         assertTrue(receivedError.get());
         assertFalse(session.isOpen());
         assertFalse(receivedReply.get());
-        verify(webSocketExceptionHandler).messageDeliveryException(notNull(), notNull());
+        verify(webSocketExceptionHandler).delegate(notNull(), notNull());
+        verify(webSocketExceptionHandler).accessDeniedException(notNull(), notNull());
     }
 
     WebSocketHandler makeWebSocketHandler(AtomicBoolean receivedReply, AtomicBoolean receivedError) {
@@ -131,7 +133,8 @@ class IntegrationWebSocketSecurityTest extends BaseWebSocketIntegrationTestRunne
         assertTrue(receivedError.get());
         assertFalse(session.isOpen());
         assertFalse(receivedReply.get());
-        verify(webSocketExceptionHandler).messageDeliveryException(notNull(), notNull());
+        verify(webSocketExceptionHandler).delegate(notNull(), notNull());
+        verify(webSocketExceptionHandler).authenticationException(notNull(), notNull());
     }
 
     /**
@@ -167,7 +170,8 @@ class IntegrationWebSocketSecurityTest extends BaseWebSocketIntegrationTestRunne
         assertFalse(session.isOpen());
         assertFalse(receivedReply.get());
 
-        verify(webSocketExceptionHandler).messageDeliveryException(notNull(), notNull());
+        verify(webSocketExceptionHandler).delegate(notNull(), notNull());
+        verify(webSocketExceptionHandler).authenticationException(notNull(), notNull());
     }
 
     /**
@@ -186,5 +190,6 @@ class IntegrationWebSocketSecurityTest extends BaseWebSocketIntegrationTestRunne
         assertTrue(session.isConnected());
         session.disconnect();
         await().atMost(OPERATION_TIMEOUT, TimeUnit.SECONDS).until(() -> !session.isConnected());
+        verify(webSocketExceptionHandler).delegate(notNull(), isNull());
     }
 }

